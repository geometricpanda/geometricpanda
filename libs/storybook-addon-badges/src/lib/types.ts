import { ComponentProps } from 'react';
import { TooltipMessage } from '@storybook/components';

export type TooltipConfig =
  | Omit<ComponentProps<typeof TooltipMessage>, 'children'>
  | string;

export type BadgeConfig = {
  title?: string;
  // @deprecated: Color has been deprecated, please use the `styles` property.
  color?: string;
  // @deprecated: Contrast has been deprecated, please use the `styles` property.
  contrast?: string;
  styles?: {
    backgroundColor?: string;
    borderColor?: string;
    borderRadius?: string;
    borderStyle?: string;
    borderWidth?: string;
    color?: string;
    fontSize?: string;
    fontFamily?: string;
    fontWeight?: string;
    lineHeight?: string | number;
    textTransform?: string;
    paddingInline?: string;
    paddingBlock?: string;
  };
  tooltip?: TooltipConfig;
};

export interface BadgesConfig {
  [id: string]: BadgeConfig;
}

export interface BadgeProps {
  badge: string;
  config: BadgeConfig;
}
