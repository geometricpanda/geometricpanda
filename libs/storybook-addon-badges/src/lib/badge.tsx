import React, { FC } from 'react';
import { TBadgesConfig } from './types';
import { getColor, getContrastColor, getTitle } from './helpers';
import styled from 'styled-components'


export const BadgeWrapper = styled.span`
  align-items: center;
  justify-content: center;
  display: flex;
  height: 100%;
  user-select: none;
`;

interface BadgeProps {
  badge: string;
  config: TBadgesConfig,
}

export const BadgeInner = styled.span<BadgeProps>`
  border: 1px solid ${({ badge, config }) => getColor(badge, config)};
  color: ${({ badge, config }) => getColor(badge, config)};
  background-color: ${({ badge, config }) => getContrastColor(badge, config)};
  padding: 2px 5px;
  border-radius: 3px;
  font-size: 0.75rem;
`;

export const Badge: FC<BadgeProps> = ({ badge, config }) => (
  <BadgeWrapper>
    <BadgeInner badge={badge} config={config}>
      {getTitle(badge, config)}
    </BadgeInner>
  </BadgeWrapper>
);

