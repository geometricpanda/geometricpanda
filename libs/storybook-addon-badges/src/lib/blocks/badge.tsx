import React, { FC } from 'react';
import { styled } from '@storybook/theming';
import { WithBadgeTooltip } from './withBadgeTooltip';
import type { BadgeProps } from '../types';
import { getBadgeCustomProperty, getTitle } from '../helpers/helpers';

export const BadgeWrapper = styled.span(({ theme }) => ({
  display: 'flex',
  paddingInline: theme.layoutMargin / 2,
}));

export const BadgeInner = styled.span<Omit<BadgeProps, 'config'>>(
  ({ badge }) => ({
    borderColor: `var(${getBadgeCustomProperty(badge, 'border-color')})`,
    borderRadius: `var(${getBadgeCustomProperty(badge, 'border-radius')})`,
    borderStyle: `var(${getBadgeCustomProperty(badge, 'border-style')})`,
    borderWidth: `var(${getBadgeCustomProperty(badge, 'border-width')})`,
    color: `var(${getBadgeCustomProperty(badge, 'color')})`,
    backgroundColor: `var(${getBadgeCustomProperty(
      badge,
      'background-color'
    )})`,
    fontSize: `var(${getBadgeCustomProperty(badge, 'font-size')})`,
    fontFamily: `var(${getBadgeCustomProperty(badge, 'font-family')})`,
    'font-weight': `var(${getBadgeCustomProperty(badge, 'font-weight')})`,
    lineHeight: `var(${getBadgeCustomProperty(badge, 'line-height')})`,
    'text-transform': `var(${getBadgeCustomProperty(badge, 'text-transform')})`,
    paddingInline: `var(${getBadgeCustomProperty(badge, 'padding-inline')})`,
    paddingBlock: `var(${getBadgeCustomProperty(badge, 'padding-block')})`,
    display: 'block',
  })
);

const BadgeInnerWithTooltip = WithBadgeTooltip(BadgeInner);

export const Badge: FC<BadgeProps> = ({ badge, config }) => (
  <BadgeWrapper>
    <BadgeInnerWithTooltip badge={badge} config={config}>
      {getTitle(badge, config)}
    </BadgeInnerWithTooltip>
  </BadgeWrapper>
);
