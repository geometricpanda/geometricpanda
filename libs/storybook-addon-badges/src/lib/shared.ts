import type { BadgeConfig } from './types';

export const ADDON_ID = '@geometricpanda/storybook-addon-badges';
export const ADDON_TITLE = 'Storybook Addon Badges';

export const PARAM_BADGES_CONFIG_KEY = 'badgesConfig';
export const PARAM_BADGES_KEY = 'badges';

export enum BADGE {
  DEFAULT = 'default',
  BETA = 'beta',
  STABLE = 'stable',
  NEEDS_REVISION = 'needs-revision',
  OBSOLETE = 'obsolete',
  EXPERIMENTAL = 'experimental',
  DEPRECATED = 'deprecated',
}

export const defaultBadgesConfig = {
  [BADGE.DEFAULT]: {
    title: 'Badge',
    contrast: '#EDEFF5',
    color: '#474D66'
  },
  [BADGE.BETA]: {
    title: 'Beta',
    contrast: '#D6E0FF',
    color: '#2952CC'
  },
  [BADGE.STABLE]: {
    title: 'Stable',
    contrast: '#DCF2EA',
    color: '#317159'
  },
  [BADGE.NEEDS_REVISION]: {
    title: 'Needs Revision',
    contrast: '#FFEFD2',
    color: '#66460D'
  },
  [BADGE.OBSOLETE]: {
    title: 'Obsolete',
    contrast: '#F9DADA',
    color: '#7D2828'
  },
  [BADGE.EXPERIMENTAL]: {
    title: 'Experimental',
    contrast: '#E7E4F9',
    color: '#6E62B6'
  },
  [BADGE.DEPRECATED]: {
    title: 'Deprecated',
    contrast: '#F8E3DA',
    color: '#85462B'
  }
};

export const defaultBadgeConfig: BadgeConfig = defaultBadgesConfig[BADGE.DEFAULT];
