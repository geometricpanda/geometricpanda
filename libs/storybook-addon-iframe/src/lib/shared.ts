export const ADDON_ID = '@geometricpanda/storybook-addon-iframe';
export const ADDON_TITLE = 'External Content';
export const PARAM_KEY = `iframe`;


export interface AddonParameters {
  timeout: number;
  url: string;
}

export const DEFAULT_PARAMETERS: AddonParameters = {
  timeout: 10000,
  url: ''
};
