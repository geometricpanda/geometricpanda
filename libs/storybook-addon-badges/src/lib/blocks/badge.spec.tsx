import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { BadgesConfig } from '../types';
import { Badge, BadgeInner } from './badge';
import { defaultBadgeConfig } from '../shared';

enum BADGE_KEYS {
  BADGE = 'badge',
  NO_TITLE = 'no-title',
  NO_COLOR = 'no-color',
  NO_CONTRAST = 'no-contrast',
  UNKNOWN_BADGE = 'unknown'
}

const badgeConfig: BadgesConfig = {
  [BADGE_KEYS.BADGE]: {
    title: 'Badge',
    contrast: '#6200EE',
    color: '#018786'
  },
  [BADGE_KEYS.NO_TITLE]: {
    contrast: '#6200EE',
    color: '#018786'
  },
  [BADGE_KEYS.NO_COLOR]: {
    title: 'Badge',
    contrast: '#6200EE'
  },
  [BADGE_KEYS.NO_CONTRAST]: {
    title: 'Badge',
    color: '#018786'
  }
};

describe('Badge', () => {

  it('should render a styled badge from config', () => {
    const tree = renderer.create(<Badge badge={BADGE_KEYS.BADGE} config={badgeConfig} />);
    expect(tree).toMatchSnapshot();
  });

  const badgeMap = [
    ['Full Badge', BADGE_KEYS.BADGE, badgeConfig[BADGE_KEYS.BADGE].title],
    ['No Title', BADGE_KEYS.NO_TITLE, BADGE_KEYS.NO_TITLE],
    ['No Color', BADGE_KEYS.NO_COLOR, badgeConfig[BADGE_KEYS.NO_COLOR].title],
    ['No Contrast', BADGE_KEYS.NO_CONTRAST, badgeConfig[BADGE_KEYS.NO_CONTRAST].title],
    ['Unknown', BADGE_KEYS.UNKNOWN_BADGE, BADGE_KEYS.UNKNOWN_BADGE]
  ];

  describe.each(badgeMap)(`Badge - %s`, (_, key, title) => {
    it(`should have a title of ${title} `, () => {
      const tree = renderer.create(<Badge badge={key} config={badgeConfig[key]} />);
      expect(JSON.stringify(tree)).toContain(title);
    });
  });

  const badgeInnerMap = [
    ['Full Badge', BADGE_KEYS.BADGE, badgeConfig[BADGE_KEYS.BADGE].contrast, badgeConfig[BADGE_KEYS.BADGE].color],
    ['No Title', BADGE_KEYS.NO_TITLE, badgeConfig[BADGE_KEYS.NO_TITLE].contrast, badgeConfig[BADGE_KEYS.NO_TITLE].color],
    ['No Color', BADGE_KEYS.NO_COLOR, badgeConfig[BADGE_KEYS.NO_COLOR].contrast, defaultBadgeConfig.color],
    ['No Contrast', BADGE_KEYS.NO_CONTRAST, defaultBadgeConfig.contrast, badgeConfig[BADGE_KEYS.NO_CONTRAST].color],
    ['Unknown', BADGE_KEYS.UNKNOWN_BADGE, defaultBadgeConfig.contrast, defaultBadgeConfig.color]
  ];

  describe.each(badgeInnerMap)('Badge Inner - %s', (_, key, contrast, color) => {

    it(`should have a color of ${color} `, () => {
      const tree = renderer.create(<BadgeInner badge={key} config={badgeConfig[key]} />);
      expect(tree.toJSON()).toHaveStyleRule('color', color);
    });

    it(`should have a contrast color of ${color} `, () => {
      const tree = renderer.create(<BadgeInner badge={key} config={badgeConfig[key]} />);
      expect(tree.toJSON()).toHaveStyleRule('background-color', contrast);
    });

  });

});
