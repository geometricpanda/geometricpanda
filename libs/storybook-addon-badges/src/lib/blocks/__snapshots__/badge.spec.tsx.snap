// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge should render a styled badge from config 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  border: 1px solid #474D66;
  color: #474D66;
  background-color: #EDEFF5;
  padding: 2px 5px;
  border-radius: 3px;
  font-size: 0.75rem;
  text-transform: uppercase;
  font-weight: bold;
}

<span
  className="c0"
>
  <span
    className="c1"
  >
    badge
  </span>
</span>
`;
